streams/src/main/java/org/apache/kafka/streams/kstream/internals/StreamsTopologyGraphImpl.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.processor.internals.InternalTopologyBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+class StreamsTopologyGraphImpl implements StreamsTopologyGraph {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StreamsTopologyGraphImpl.class);
+    private StreamGraphNode previousNode;
+    private final AtomicInteger nodeIdCounter = new AtomicInteger(0);
+    public static final String TOPOLOGY_ROOT = "root";
+    protected final StreamGraphNode root = new StreamGraphNode(TOPOLOGY_ROOT, TOPOLOGY_ROOT, false) {
+        @Override
+        void writeToTopology(InternalTopologyBuilder topologyBuilder) {
+            // no-op for root node
+        }
+    };
+
+    private final Map<StreamGraphNode, Set<StreamGraphNode>> repartitioningNodeToRepartitioned = new HashMap<>();
+    private final Map<StreamGraphNode, StreamGraphNode> stateStoreNodeToSinkNodes = new HashMap<>();
+    private final Map<String, StreamGraphNode> nameToGraphNode = new HashMap<>();
+
+    StreamsTopologyGraphImpl() {
+        nameToGraphNode.put(TOPOLOGY_ROOT, root);
+    }
+
+
+    @Override
+    public <K, V> void addNode(final StreamGraphNode<K, V> node) {
+        node.setId(nodeIdCounter.getAndIncrement());
+
+        if (node.parentNodeName() == null) {
+            LOG.warn("Updating node {} with predecessor name {}", node, previousNode.nodeName());
```

Comment:
Should this really be `WARN`? I think this is the most common case?

Reply:
When building the nodes we should always have access to the parent name (the calling KStream/KTable instance) so ideally, no node should get added without a parent name.

Reply:
Ack.